ampd/src/stacks/contract_verifier.rs

```diff
@@ -0,0 +1,249 @@
+use clarity::vm::types::{
+    BufferLength, PrincipalData, SequenceSubtype, TupleTypeSignature, TypeSignature,
+};
+use clarity::vm::{ClarityName, Value};
+
+use crate::stacks::error::Error;
+use crate::stacks::http_client::{Client, TransactionEvents};
+
+pub fn get_verify_contract_params(
+    event: &TransactionEvents,
+) -> Option<(PrincipalData, PrincipalData)> {
+    let payload = get_payload_from_contract_call_event(event).ok()?;
+
+    let inner_verify_contract_type_signature = TupleTypeSignature::try_from(vec![
+        (
+            ClarityName::from("reference-address"),
+            TypeSignature::PrincipalType,
+        ),
+        (
+            ClarityName::from("contract-address"),
+            TypeSignature::PrincipalType,
+        ),
+    ])
+    .ok()?;
+
+    let verify_contract_type_signature = TupleTypeSignature::try_from(vec![(
+        ClarityName::from("verify-contract"),
+        TypeSignature::TupleType(inner_verify_contract_type_signature),
+    )])
+    .ok()?;
+
+    let mut verify_contract = Value::try_deserialize_bytes(
+        &payload,
+        &TypeSignature::TupleType(verify_contract_type_signature),
+        true,
+    )
+    .ok()?
+    .expect_tuple()
+    .ok()?
+    .get_owned("verify-contract")
+    .ok()?
+    .expect_tuple()
+    .ok()?;
+
+    let reference_address = verify_contract
+        .data_map
+        .remove("reference-address")
+        .ok_or(Error::PropertyEmpty)
+        .ok()?
+        .expect_principal()
+        .ok()?;
+    let contract_address = verify_contract
+        .get_owned("contract-address")
+        .ok()?
+        .expect_principal()
+        .ok()?;
+
+    Some((reference_address, contract_address))
+}
+
+pub async fn verify_contract_code(
+    http_client: &Client,
+    reference_address: PrincipalData,
+    contract_address: PrincipalData,
+) -> Result<bool, Box<dyn std::error::Error>> {
+    let reference_contract_code = http_client
+        .get_contract_info(format!("{}", reference_address.to_string()).as_str())
+        .await?;
+
+    let actual_contract_code = http_client
+        .get_contract_info(format!("{}", contract_address.to_string()).as_str())
+        .await?;
+
+    Ok(reference_contract_code.source_code == actual_contract_code.source_code)
+}
+
```

Comment:
I am worried about the performance implications of this. As far as I can tell, this is downloading two sets of source code and then comparing them? Is it comparing a hash of the source code or the full bytes of the source code?

Reply:
No longer applicable